### PR TITLE
Add diagnostics callback for tlscommon configs and httptransport

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Add diagnostics callbacks for TLS configuration to make it easier to
include in diagnostics bundles. Diagnostics info will geneally verify
that the certs or cas are able to load, and display some basic info such
as cert names, expiry times, fingerprints, etc. httptransport callback
will attempt an HTTP request and return information collected from
httptrace about the request.

## Why is it important?

Allows more information about TLS to be present in diagnostics to help troubleshoot any certificate issues.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by michel-laterman** [comment]
Example of the hooks functions and outputs added to a (draft) fleet-server pr: https://github.com/elastic/fleet-server/pull/3587

File: transport/httpcommon/diag.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			logger.Printf("Request %d to %s starting", i, req.URL.String())
+			req = req.WithContext(httptrace.WithClientTrace(req.Context(), ct))
+			if resp, err := rt.RoundTrip(req); err != nil {
+				logger.Printf("request %d error: %v", i, err)
+			} else {
+				resp.Body.Close()
+				logger.Printf("request %d successful. status=%d", i, resp.StatusCode)
+			}
+		}
+		return b.Bytes()
```
Comment (michel-laterman) [inline transport/httpcommon/diag.go:108]: As seen in the demo fleet-server PR the output would look like:

```
2024/05/29 00:18:41.571489 No TLS settings
2024/05/29 00:18:41.571492 Proxy disable=false url=<nil>
2024/05/29 00:18:41.571541 Request 0 to https://1df578882a2d4d0190aabbbe4ac77268.us-central1.gcp.qa.cld.elstc.co:443 starting
2024/05/29 00:18:41.571576 GetConn called for "1df578882a2d4d0190aabbbe4ac77268.us-central1.gcp.qa.cld.elstc.co:443"
2024/05/29 00:18:41.710378 Connection started to tcp:34.71.9.202:443
2024/05/29 00:18:41.834875 Connection to tcp:34.71.9.202:443 done, err: <nil>
2024/05/29 00:18:41.984139 TLS handshake starting
2024/05/29 00:18:41.984148 TLS handshake done. state={Version:772 HandshakeComplete:true DidResume:false CipherSuite:4865 NegotiatedProtocol: NegotiatedProtocolIsMutual:true ServerName:1df578882a2d4d0190aabbbe4ac77268.us-central1.gcp.qa.cld.elstc.co PeerCertificates:[0xc00021b600 0xc00021bb80 0xc00021c100] VerifiedChains:[] SignedCertificateTimestamps:[] OCSPResponse:[] TLSUnique:[] ekm:0x5557d33c1f20} err=<nil>
2024/05/29 00:18:41.984267 GotConn for "34.71.9.202:443"
2024/05/29 00:18:41.984371 Wrote request headers
2024/05/29 00:18:41.984385 Wrote request err=<nil>
2024/05/29 00:18:42.111030 Response started
2024/05/29 00:18:42.111184 request 0 successful. status=401
```
The goal of this diagnostics is to make sure that we are able to connect to the endpoint, not what that endpoint actually returns (status code is included for information, in my example i didn't set the auth header)

File: transport/tlscommon/diag.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		logger := log.New(&b, "tlscommon.ServerConfig: ", 0)
+		logger.Printf("Start diagnostics %s", time.Now().UTC())
+		logger.Printf("verification_mode=%s", c.VerificationMode)
+		logger.Printf("client_auth=%s", c.ClientAuth)
+		logger.Printf("ca_sha256=%v", c.CASha256)
+
+		diagCertificate(logger, &c.Certificate)
+		diagCAs(logger, c.CAs)
+
+		return b.Bytes()
```
Comment (michel-laterman) [inline transport/tlscommon/diag.go:72]: From the fleet-server demo, output is:

```
tlscommon.ServerConfig: Start diagnostics 2024-05-29 00:18:41.569890423 +0000 UTC
tlscommon.ServerConfig: verification_mode=full
tlscommon.ServerConfig: client_auth=none
tlscommon.ServerConfig: ca_sha256=[]
tlscommon.ServerConfig: CertificateSettings: checking certificate keypair
tlscommon.ServerConfig: CertificateSettings: certificate keypair OK.
tlscommon.ServerConfig: CertificateSettings: cert 0 - Subject: CN=fleet-server-dev,O=elastic-fleet
	Issuer: CN=localhost,O=elastic-fleet
	NotBefore: 2024-05-29 00:18:15 +0000 UTC
	NotAfter: 2034-05-29 00:18:15 +0000 UTC
	Fingerprint: EvAr8erNJZJ0BoXor1jnCl5X2WCU79ISMgwMITOEJBs=
	SAN IP: []
	SAN DNS: [fleet-server-dev]
tlscommon.ServerConfig: CertificateAuthorities: certificate_authorities provided.
tlscommon.ServerConfig: CertificateAuthorities: - cert 0 Subject: CN=localhost,O=elastic-fleet
	IsCa: true
	BasicConstraintsValid: true
	NotBefore: 2024-05-29 00:18:15 +0000 UTC
	NotAfter: 2034-05-29 00:18:15 +0000 UTC
	Fingerprint: cSawoNEa2cSPV4Mwce4mRysrcBGtHBGcSeoBL+Cr48Q=
```

We want to make sure that certificate & ca information is provided in the diagnostics

**Review by michel-laterman — COMMENTED** [review]
I would like feedback, specifically if we need more information to help troubleshoot issues.

File: transport/httpcommon/diag.go (hunk 1 above)
Comment (leehinman) [reply transport/httpcommon/diag.go:108]: Can you get any of the cert info that is pointed to by `PeerCertificates:[0xc00021b600 0xc00021bb80 0xc00021c100]` in the example?  similar to what you have cert & ca in the next example.

**Comment by leehinman** [comment]
> I would like feedback, specifically if we need more information to help troubleshoot issues.

This may or may not be helpful, so feel free to disregard.

I'd like to see some failure scenarios, paired with the new diagnostics showing how they give the info needed to correctly diagnose the problem.  I'm thinking of things like:

1. User configures NonTLS client but connects to TLS server
2. User configures TLS client but connects to NonTLS server
3. User configures TLS client, no client cert, private key but connects to TLS server that requires TLS client authentication
4. User configures TLS client but connects to server that has expired certificate
5. User configures TLS client but doesn't have CA that signed TLS server cert in it's trust store
6. User configures TLS client but server requires an algorithm that isn't supported by the client
7. User configures TLS client but connects to Proxy that doesn't have the server name as a SAN

File: transport/httpcommon/diag.go (hunk 1 above)
Comment (michel-laterman) [reply transport/httpcommon/diag.go:121]: Forgive the typo
```
2024/05/29 22:10:06.844374 No TLS settings
2024/05/29 22:10:06.844426 Proxy disable=false url=<nil>
2024/05/29 22:10:06.844562 Request 0 to https://3c965949205c4e3cb943d1765bb4ced0.us-central1.gcp.qa.cld.elstc.co:443 starting
2024/05/29 22:10:06.844693 GetConn called for "3c965949205c4e3cb943d1765bb4ced0.us-central1.gcp.qa.cld.elstc.co:443"
2024/05/29 22:10:06.926192 Connection started to tcp:34.71.9.202:443
2024/05/29 22:10:07.044556 Connection to tcp:34.71.9.202:443 done, err: <nil>
2024/05/29 22:10:07.222023 TLS handshake starting
2024/05/29 22:10:07.222056 TLS handshake done. state={Version:772 HandshakeComplete:true DidResume:false CipherSuite:4865 NegotiatedProtocol: NegotiatedProtocolIsMutual:true ServerName:3c965949205c4e3cb943d1765bb4ced0.us-central1.gcp.qa.cld.elstc.co PeerCertificates:[0xc0004b6b00 0xc0004b7080 0xc0004b7600] VerifiedChains:[] SignedCertificateTimestamps:[] OCSPResponse:[] TLSUnique:[] ekm:0x561fc2929f20} err=<nil>
2024/05/29 22:10:07.222167 Peer certificate count 3
2024/05/29 22:10:07.222392 - Peer Certificate 0
	Subject=CN=*.us-central1.gcp.qa.cld.elstc.co
	Issuer=CN=R3,O=Let's Encrypt,C=US
	IsCA=false
	BasicConstrintsValid=true
	NotBefore=2024-05-05 23:35:35 +0000 UTC
	NotAfter=2024-08-03 23:35:34 +0000 UTC
	Fingerprint=dr41bmmfovR/W9yZCK9EZ00RbjlwcVc2ivxaaA3FkfI=
	SAN IP=[]
	SAN DNS=[*.apm.psc.us-central1.gcp.qa.cld.elstc.co *.apm.us-central1.gcp.qa.cld.elstc.co *.app-search.psc.us-central1.gcp.qa.cld.elstc.co *.app-search.us-central1.gcp.qa.cld.elstc.co *.ent-search.psc.us-central1.gcp.qa.cld.elstc.co *.ent-search.us-central1.gcp.qa.cld.elstc.co *.ent.psc.us-central1.gcp.qa.cld.elstc.co *.ent.us-central1.gcp.qa.cld.elstc.co *.es.psc.us-central1.gcp.qa.cld.elstc.co *.es.us-central1.gcp.qa.cld.elstc.co *.fleet.psc.us-central1.gcp.qa.cld.elstc.co *.fleet.us-central1.gcp.qa.cld.elstc.co *.kb.psc.us-central1.gcp.qa.cld.elstc.co *.kb.us-central1.gcp.qa.cld.elstc.co *.profiling.psc.us-central1.gcp.qa.cld.elstc.co *.profiling.us-central1.gcp.qa.cld.elstc.co *.psc.us-central1.gcp.qa.cld.elstc.co *.symbols.psc.us-central1.gcp.qa.cld.elstc.co *.symbols.us-central1.gcp.qa.cld.elstc.co *.us-central1.gcp.qa.cld.elstc.co]
	SAN URI=[]
2024/05/29 22:10:07.223034 - Peer Certificate 1
	Subject=CN=R3,O=Let's Encrypt,C=US
	Issuer=CN=ISRG Root X1,O=Internet Security Research Group,C=US
	IsCA=true
	BasicConstrintsValid=true
	NotBefore=2020-09-04 00:00:00 +0000 UTC
	NotAfter=2025-09-15 16:00:00 +0000 UTC
	Fingerprint=jQJTbIh0grw0/1TkHSumWb+Fs0Ggogr621gT3PvPKG0=
	SAN IP=[]
	SAN DNS=[]
	SAN URI=[]
2024/05/29 22:10:07.223097 - Peer Certificate 2
	Subject=CN=ISRG Root X1,O=Internet Security Research Group,C=US
	Issuer=CN=ISRG Root X1,O=Internet Security Research Group,C=US
	IsCA=true
	BasicConstrintsValid=true
	NotBefore=2015-06-04 11:04:38 +0000 UTC
	NotAfter=2035-06-04 11:04:38 +0000 UTC
	Fingerprint=C5+lpZ7tcVwmwQIMcRtPbsQtWLABXhQzejna0wHFr8M=
	SAN IP=[]
	SAN DNS=[]
	SAN URI=[]
2024/05/29 22:10:07.223100 Verified chains count: 0
2024/05/29 22:10:07.223565 GotConn for "34.71.9.202:443"
2024/05/29 22:10:07.224495 Wrote request headers
2024/05/29 22:10:07.224507 Wrote request err=<nil>
2024/05/29 22:10:07.332675 Response started
2024/05/29 22:10:07.333657 request 0 successful. status=401
```

File: transport/tlscommon/diag.go (hunk 2 above)
Comment (michel-laterman) [reply transport/tlscommon/diag.go:72]: Slight update after I added `CertDiagString`
```
tlscommon.ServerConfig: Start diagnostics 2024-05-29 22:10:07.333735518 +0000 UTC
tlscommon.ServerConfig: verification_mode=full
tlscommon.ServerConfig: client_auth=none
tlscommon.ServerConfig: ca_sha256=[]
tlscommon.ServerConfig: CertificateSettings: checking certificate keypair
tlscommon.ServerConfig: CertificateSettings: certificate keypair OK.
tlscommon.ServerConfig: CertificateSettings: cert 0 Subject=CN=fleet-server-dev,O=elastic-fleet
	Issuer=CN=localhost,O=elastic-fleet
	IsCA=false
	BasicConstrintsValid=false
	NotBefore=2024-05-29 22:08:32 +0000 UTC
	NotAfter=2034-05-29 22:08:32 +0000 UTC
	Fingerprint=w04CVUAd7Tfnd/AZ/A8YB9bCUAImGeGXBAUPnhEvGgM=
	SAN IP=[]
	SAN DNS=[fleet-server-dev]
	SAN URI=[]
tlscommon.ServerConfig: CertificateAuthorities: certificate_authorities provided.
tlscommon.ServerConfig: CertificateAuthorities: - cert 0 Subject=CN=localhost,O=elastic-fleet
	Issuer=CN=localhost,O=elastic-fleet
	IsCA=true
	BasicConstrintsValid=true
	NotBefore=2024-05-29 22:08:23 +0000 UTC
	NotAfter=2034-05-29 22:08:23 +0000 UTC
	Fingerprint=QHVJrVpzEHBJTTr6F9RW6TnYbj9ckPD2BQ5JphTzh7M=
	SAN IP=[]
	SAN DNS=[localhost]
	SAN URI=[]
```

File: transport/httpcommon/diag.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	}
+}
+
+// isGoHTTPResp detects if the response is one that a go http.Server sends if an HTTP request is made to an HTTPS server.
+// non Go servers may return a net.OpError instead.
+func isGoHTTPResp(r *http.Response) bool {
```
Comment (michel-laterman) [inline transport/httpcommon/diag.go:136]: This behaviour is a part of go servers, and also occurs for our current cloud deplyments

File: transport/httpcommon/diag.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	p, err := io.ReadAll(r.Body)
+	if err != nil {
+		return false
+	}
+	return string(p) == "Client sent an HTTP request to an HTTPS server.\n"
+}
+
+// diagError tries to diagnose the error and return a cause/possible cause in a human readable format.
+// If no matching errors are found err.Error is returned.
+func diagError(err error) string {
```
Comment (michel-laterman) [inline transport/httpcommon/diag.go:149]: @leehinman, I've tried to add this to explain errors based on the scenarios you provided (with test cases for most of them). we can add more in the future as well

File: transport/httpcommon/diag.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+		}
+	}
+	return func() []byte {
+		var b bytes.Buffer
+		rt, err := settings.RoundTripper(opts...)
+		if err != nil {
+			b.WriteString("unable to get roundtripper: " + err.Error())
+			return b.Bytes()
+		}
+		logger := log.New(&b, "", log.LstdFlags|log.Lmicroseconds|log.LUTC)
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply transport/httpcommon/diag.go:56]: ease of implementation.

File: transport/httpcommon/diag_test.go [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+		require.NoError(t, err)
+		defer resp.Body.Close()
+		require.False(t, isGoHTTPResp(resp), "expected go https response on https server")
+	})
+}
+
+func Test_HTTPRequestOnHTTPSPort(t *testing.T) {
+	// This checks if HTTP was use used to communicate with an HTTPS server resulting in an error
+	// isGoHTTPResp does the same but requires that the responding server is a go http.Server
+	t.Skip("test used to validate behaviour of http.Client on external servers")
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply transport/httpcommon/diag_test.go:144]: this was more of a sanity test, and i really don't like having our unit tests rely on an external server

**Comment by michel-laterman** [comment]
@lucabelluccini is there anything I should add/change to help with your troubleshooting efforts?

**Review by belimawr — APPROVED** [review]

**Review by leehinman — APPROVED** [review]
LGTM.  Thanks for adding the test cases.

**Comment by lucabelluccini** [comment]
Hello, thanks for the heads up on this one and thank you for those efforts into troubleshooting TLS.
I have few questions:
1. We will have 1  TLS diagnostic per output (e.g. the integrations output, the monitoring output, fleet server connection)? How will it work?
2. We will have a documentation section to cover some example failures a user might encounter?

Example about (1).
Imagine a user has 1 Elastic Agent.
The Elastic Agent also runs the System integration.
The EA sends logs & metrics to Elasticsearch.

So we have:
- connection to ES for monitoring (logs & metrics of EA)
- connection to ES for integrations (system integration)
- connection to Fleet Server (Fleet Server enroll/checkins)

We will get 3 TLS diags?

**Comment by michel-laterman** [comment]
@lucabelluccini
1. yes, we'll have a request trace file per destination type; these will need to be manually added in future PRs. The implemenation issue for the elastic-agent is: https://github.com/elastic/elastic-agent/issues/4880.
  just note the fleet-server implementation: https://github.com/elastic/fleet-server/pull/3587 currently only tests the connection to ES; it does not check for monitoring output; but I can add that check in the pr
2. I added `diagError` to try to turn an error that occurs in the connection into a human-readable cause such as `%v: possible cause: HTTP schema used for HTTPS server.`, or `%v: caused by no trusted client CA.` but we can make a KB article to help track the causes of (connection) failures

**Comment by lucabelluccini** [comment]
Lovely - Thank you @michel-laterman & team ❤️

**Pull request merged**